### ISSUE TITLE
[BUG] Restore variables values after resume from a bookmark.

### ISSUE THREAD

**Issue by edward-yuen-tfs** [elsa-workflows/elsa-core#5547]
## Description
For a variable defined in the parent workflow, create an activity to decrement the variable and creates a bookmark afterwards.  After the bookmark is resumed, the variable value is not restored properly in the parent workflow and the variable value continues to stay as default.

## Steps to Reproduce
To help us identify the issue more quickly, please follow these guidelines:

1. Create a sample workflow like below.
 public class CountDownWorkflow : WorkflowBase
 {
     protected override void Build(IWorkflowBuilder builder)
     {
         var numberOfItems = builder.WithVariable("NumberOfItems", 5).WithWorkflowStorage();
  
         builder.Root = new Sequence
         {
             Variables = { numberOfItems },
             Activities =
             {
             new SetVariable<int>(numberOfItems, 5),  // Try to explicit set here.
             new WriteLine(context => "Initial number of items = " + numberOfItems.Get(context) ),
             new While(context => numberOfItems.Get(context) > 0)
             {
                 Body = new Sequence
                 {
                     Activities =
                     {
                         // Increment counter variable.
                         new CustomActivity(),
                         new WriteLine(context => "Number of items = " + numberOfItems.Get(context) ),
                     }
                 }
             }
         }
         };
     }
 }

2. The CustomActivity code is like below.
 public class CustomActivity : CodeActivity<int>
 {
     public CustomActivity()
     {
     }
     protected override void Execute(ActivityExecutionContext context)
     {
         var numberOfItems = context.GetVariable<int>("NumberOfItems");
         context.SetVariable("NumberOfItems", numberOfItems - 1);
         context?.CreateBookmark();
     }
 }

3. Trigger the workflow.  Jot down the instance Id.
4. Call resume multiple times.   Observed that the value does not decrement further in the While loop.
5. You can call the following 2 apis to try to test it with the attached sample program,
https://localhost:5001/testapi/submit   - jot down the instance id.
https://localhost:5001/testapi/resume - put the instance id in the body.

## Expected Behavior
Variable is decremented and the parent workflow can print it until the variable reaches zero and exit the While loop

## Actual Behavior
The variable value stays as 4 all the time.   Do not decrement to zero.

## Screenshots
![image](https://github.com/elsa-workflows/elsa-core/assets/54376892/a7132830-4b1d-47b6-89a1-468b67d447da)

## Related Issues
https://github.com/elsa-workflows/elsa-core/issues/4148

[ResumeWorkflowApp.Web.zip](https://github.com/user-attachments/files/15686417/ResumeWorkflowApp.Web.zip)

**Comment by sfmskywalker** [comment]
Thanks @edward-yuen-tfs for the level of detail + working repro - I can indeed reproduce the issue.

**Comment by sfmskywalker** [comment]
I figured out what the problem is. Actually, there are two issues:

**1. Variable Persistence**

Notice the following line in your workflow:

<img width="945" alt="image" src="https://github.com/elsa-workflows/elsa-core/assets/938393/99934ab7-bfe7-4702-b922-54cc9667a5a3">

When using `builder.WithVariable()`, it will create a variable on the workflow level.
But when you assign a variable with the same name on a child activity, such as `Sequence` in your workflow, that variable becomes closer to your custom activity, which means that your custom activity will be setting the variable's value in the scope of the `Sequence` activity and not in the scope of the workflow.

That on its own wouldn't be a big deal, if it weren't for the fact that as of this writing, **only workflow-level variables are persisted**.

Given that the workflow level variable is never updated, it explains that its value is lost after resuming the workflow instance.

So, if you remove that line that I enrectangled red in the screenshot, it will work - _almost_.

**2. Variable Scoping Bug**

While troubleshooting this issue, I encountered a bug that causes your workflow to print the same number twice (the number 4, as you will see).
This is due to a scoping inconsistency bug in Elsa. I am already working on a fix that I will push out over the weekend, which means that you'll be able to take advantage by referencing the preview packages once they are built.

I'll update progress here soon.

**Comment by edward-yuen-tfs** [comment]
Thank you for looking into that.   Just want to be confirm on this.   

1. "**only workflow-level variables are persisted**."  - that means 
a. activity level variables are not persisted.
b. For the picture below, which I have commented out the line you suggest.
![image](https://github.com/elsa-workflows/elsa-core/assets/54376892/49009ee6-8cf8-4998-a45c-ab7888092a9c)
- As I already set the value in the orange box, I do not need to set the value in the red box.   Both lines in fact refer to the same scope.
- I do not need the commented line as it shall be considered as defining a new variable within the sequence scope.   It is not considered as passing the workflow scope variable to the sequence scope (it is not needed.)
c. After I commented the line, CustomActivity is actually setting the variable value to the workflow scope.

Just want to confirm if my understanding is correct.
